### PR TITLE
Update expected errors

### PR DESCRIPTION
**Pull request by malwaregarry**
This PR simply decouples the definition of error strings and regexes with adding them to  ExpectedErrors.

### CODE REVIEW

**Comment by mrigger** [comment]
Some more thoughts:
* We could also implement a static factory method `ExpectedErrors.of`, similar to `List.of`.
* Since we don't want to have duplicate errors, we should actually use a `Set` rather than a `List` (I see that the `ExpectedErrors` class already does that).

**Comment by malwaregarry** [comment]
@mrigger hmm, there is a static  factory method `from` in ExpectedErrors. Do I stick with that naming or should I rename it?

**Comment by mrigger** [comment]
I think either would work, as you prefer.

File: src/sqlancer/common/query/ExpectedErrors.java
Comment (mrigger) [inline src/sqlancer/common/query/ExpectedErrors.java:59]: Btw, I think we could also provide convenience classes that make adding regular expressions less tedious (e.g., `fromRegexes(String... errors)` that takes the strings and instantiates `Pattern` objects.

**Comment by malwaregarry** [comment]
I added a new ExpectedErrorBuilder class to handle the creation of new errors. I kept the original functions because I think some dependencies require the fact that ExpectedErrors is mutable. 

> we don't want to have duplicate errors, we should actually use a Set rather than a List (I see that the ExpectedErrors class already does that).

I changed Regexes to use a set as well. 

I think this PR can be merged if you think it's good.

**Review by mrigger — APPROVED** [review]

**Comment by mrigger** [comment]
> Edit: just saw the comment. Don't merge yet

Okay, let me know when.

It would btw be nice if we could also subsequently add some unit tests to test such core functionality.

**Comment by malwaregarry** [comment]
@mrigger Clickhouse is now failing. I checked that it doesn't use the methods that were changed, and doesn't use regexes, so the  change to a hashset isn't the problem. And since the commit 3 days back passed the CI, it shouldn't have anything to do with this PR.

I just checked that there was an update to the clickhouse docker image 5 hours ago, so it could be anew bugs.

https://hub.docker.com/r/clickhouse/clickhouse-server/tags

**Comment by mrigger** [comment]
It seems also PostgreSQL started failing, which I think still worked before. Do you think this is also unrelated?

**Comment by mrigger** [comment]
The ClickHouse "java.io.IOException: std::exception. Code: 1001, type: std::bad_variant_access, e.what() = bad_variant_access (version 24.3.1.1823 (official build))" issue seems like a potential bug. @qoega, any input on this?

**Comment by malwaregarry** [comment]
hmm, in that case I think i messed up somewhere. let me run locally.

**Comment by qoega** [comment]
I've seen it our ci recently. It is already fixed!

https://github.com/ClickHouse/ClickHouse/pull/61654

**Comment by malwaregarry** [comment]
The checks pass now. 

Postgres was due to an error on my side. Thanks for spotting it.

File: src/sqlancer/clickhouse/ClickHouseErrors.java
Comment (mrigger) [inline src/sqlancer/clickhouse/ClickHouseErrors.java:43]: Just noticed. Some of these are known bugs, for which we should the pattern explained at https://github.com/sqlancer/sqlancer/blob/main/CONTRIBUTING.md#unfixed-bugs (e.g., see the Citus implementation). I guess it's not in scope of the PR, but if you are interested, @malwaregarry, you could address those (perhaps some of them have also been fixed by now).

File: src/sqlancer/sqlite3/SQLite3Errors.java
Comment (mrigger) [inline src/sqlancer/sqlite3/SQLite3Errors.java:73]: For the set operators, window functions, INDEXED by, etc: these are actually not expression errors, so we'd like to add those only if the statement might use them (adding expected errors that are not relevant might otherwise shadow some bugs). Where did you extract those?

Comment (malwaregarry) [reply src/sqlancer/sqlite3/SQLite3Errors.java:73]: I got them from the `addExpectedExpressionErrors()` function below. Just realised i forgot to update `addExpectedExpressionErrors()` to call this function

Comment (mrigger) [reply src/sqlancer/sqlite3/SQLite3Errors.java:73]: Okay, I see. After remove this from the old one, we can merge the PR.

File: src/sqlancer/clickhouse/ClickHouseErrors.java
Comment (malwaregarry) [reply src/sqlancer/clickhouse/ClickHouseErrors.java:43]: thanks for pointing this out. I can work on this after I'm done with NoREC

File: src/sqlancer/sqlite3/SQLite3Errors.java
Comment (malwaregarry) [reply src/sqlancer/sqlite3/SQLite3Errors.java]: @mrigger I made the change and checks pass.

**Review by mrigger — APPROVED** [review]

**Pull request merged**
